utils/rate_limiter_test.go

```diff
@@ -15,61 +15,91 @@
+	rl.lastTick = ts
+	rl.timeNow = func() time.Time {
 		return ts
 	}
-	assert.True(t, limiter.CheckCredit(1.0))
-	assert.True(t, limiter.CheckCredit(1.0))
-	assert.False(t, limiter.CheckCredit(1.0))
+	assert.True(t, rl.CheckCredit(1.0))
+	assert.True(t, rl.CheckCredit(1.0))
+	assert.False(t, rl.CheckCredit(1.0))
 	// move time 250ms forward, not enough credits to pay for 1.0 item
-	limiter.(*rateLimiter).timeNow = func() time.Time {
+	rl.timeNow = func() time.Time {
 		return ts.Add(time.Second / 4)
 	}
-	assert.False(t, limiter.CheckCredit(1.0))
+	assert.False(t, rl.CheckCredit(1.0))
 	// move time 500ms forward, now enough credits to pay for 1.0 item
-	limiter.(*rateLimiter).timeNow = func() time.Time {
+	rl.timeNow = func() time.Time {
 		return ts.Add(time.Second/4 + time.Second/2)
 	}
-	assert.True(t, limiter.CheckCredit(1.0))
-	assert.False(t, limiter.CheckCredit(1.0))
+	assert.True(t, rl.CheckCredit(1.0))
+	assert.False(t, rl.CheckCredit(1.0))
 	// move time 5s forward, enough to accumulate credits for 10 messages, but it should still be capped at 2
-	limiter.(*rateLimiter).lastTick = ts
-	limiter.(*rateLimiter).timeNow = func() time.Time {
+	rl.lastTick = ts
+	rl.timeNow = func() time.Time {
 		return ts.Add(5 * time.Second)
 	}
-	assert.True(t, limiter.CheckCredit(1.0))
-	assert.True(t, limiter.CheckCredit(1.0))
-	assert.False(t, limiter.CheckCredit(1.0))
-	assert.False(t, limiter.CheckCredit(1.0))
-	assert.False(t, limiter.CheckCredit(1.0))
+	assert.True(t, rl.CheckCredit(1.0))
+	assert.True(t, rl.CheckCredit(1.0))
+	assert.False(t, rl.CheckCredit(1.0))
+	assert.False(t, rl.CheckCredit(1.0))
+	assert.False(t, rl.CheckCredit(1.0))
 }
 
-func TestMaxBalance(t *testing.T) {
-	limiter := NewRateLimiter(0.1, 1.0)
+func TestRateLimiterMaxBalance(t *testing.T) {
+	rl := NewRateLimiter(0.1, 1.0)
 	// stop time
 	ts := time.Now()
-	limiter.(*rateLimiter).lastTick = ts
-	limiter.(*rateLimiter).timeNow = func() time.Time {
+	rl.lastTick = ts
+	rl.timeNow = func() time.Time {
 		return ts
 	}
-	// on initialization, should have enough credits for 1 message
-	assert.True(t, limiter.CheckCredit(1.0))
+	assert.True(t, rl.CheckCredit(1.0), "on initialization, should have enough credits for 1 message")
 
 	// move time 20s forward, enough to accumulate credits for 2 messages, but it should still be capped at 1
-	limiter.(*rateLimiter).timeNow = func() time.Time {
+	rl.timeNow = func() time.Time {
 		return ts.Add(time.Second * 20)
 	}
-	assert.True(t, limiter.CheckCredit(1.0))
-	assert.False(t, limiter.CheckCredit(1.0))
+	assert.True(t, rl.CheckCredit(1.0))
+	assert.False(t, rl.CheckCredit(1.0))
+}
+
+func TestRateLimiterReconfigure(t *testing.T) {
+	rl := NewRateLimiter(1, 1.0)
+	assertBalance := func(expected float64) {
+		const precision = 0.0000001 // just some precision for comparing floats
+		delta := math.Abs(rl.balance - expected)
+		assert.Greater(t, precision, delta, "expect balance %v, was %v", expected, rl.balance)
```

Comment:
Couldn't [assert.InEpsilon ](https://godoc.org/github.com/stretchr/testify/assert#InEpsilon) work here?

Reply:
indeed, I didn't find it

Reply:
it doesn't work with 0 values, since it compares relative change

Reply:
used InDelta